plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -56,21 +62,35 @@ public class QueryInsightsService extends AbstractLifecycleComponent {
      */
     protected volatile Scheduler.Cancellable scheduledFuture;
 
+    /**
+     * Query Insights exporter factory
+     */
+    final QueryInsightsExporterFactory queryInsightsExporterFactory;
+
     /**
      * Constructor of the QueryInsightsService
      *
-     * @param threadPool     The OpenSearch thread pool to run async tasks
+     * @param clusterSettings OpenSearch cluster level settings
+     * @param threadPool The OpenSearch thread pool to run async tasks
+     * @param client OS client
      */
     @Inject
-    public QueryInsightsService(final ThreadPool threadPool) {
+    public QueryInsightsService(final ClusterSettings clusterSettings, final ThreadPool threadPool, final Client client) {
         enableCollect = new HashMap<>();
         queryRecordsQueue = new LinkedBlockingQueue<>(QueryInsightsSettings.QUERY_RECORD_QUEUE_CAPACITY);
+        this.threadPool = threadPool;
+        this.queryInsightsExporterFactory = new QueryInsightsExporterFactory(client);
+        // initialize top n queries services and configurations consumers
         topQueriesServices = new HashMap<>();
         for (MetricType metricType : MetricType.allMetricTypes()) {
             enableCollect.put(metricType, false);
-            topQueriesServices.put(metricType, new TopQueriesService(metricType));
+            topQueriesServices.put(metricType, new TopQueriesService(metricType, threadPool, queryInsightsExporterFactory));
         }
-        this.threadPool = threadPool;
+        clusterSettings.addSettingsUpdateConsumer(
+            TOP_N_LATENCY_EXPORTER_SETTINGS,
+            (settings -> getTopQueriesService(MetricType.LATENCY).setExporter(settings)),
+            (settings -> getTopQueriesService(MetricType.LATENCY).validateExporterConfig(settings))
+        );
```

Comment:
If I keep toggling the setting, we keep creating new exporter. And, we don't have reference to the original exporter. Why not create these exporters as singleton objects and allow configuring the parameters directly like indexPattern?

Reply:
IMO since different services can use the same type of exporter, using singleton means we need to maintain a map from service to exporter(s) or exporter configs, also one service could possibly support multiple exporters in the future, to avoid over engineer I think making the exporters stateless would be easier. 

But you are right, let me at least refactor this to reuse the same exporter if the type is the same as before, so that we won't create same new exporters over and over again :)